### ISSUE TITLE
enums with arbitrary values

### ISSUE THREAD

**Issue by coderextreme** [json-schema-org/json-schema-spec#665]
We'd like enums with arbitrary values.  I don't know as JSON schema supports this easily (think multiple enums with oneOf's), except as arbitrary values (type string).   The enum's non arbitrary values provide useful "user interface".  In other words, we'd like a pull down list with some default values and a type-in (but we can drive this user interface from somewhere else...it's not necessary in JSON schema.   But a reasonable error report from validation suggesting possible values would be welcome.

I'm currently using draft-07.  I don't really have a solution, except perhaps change enum to "optional_values" or "default_values" when this occurs.  I don't really have good names either.

I'm waiting for some god to pronounce the solution to my problems already exists in draft-07.

I'm starting to think our object model will provide an adequate schema, since our schema is derived from it!

Thanks,

John

**Comment by gregsdennis** [comment]
To be clear, are you looking for:

1. only `string` values, but where the user can add some values, or
2. any kind of value (`string`, `number`, `object`, etc.) as a member of an enum?

I'm guessing (1) from your description, but from "arbitrary values" I can rule out (2).

If (1), then it seems that you want to be able to allow custom values, in which case, I'd steer clear of `enum`.  Just use `"type":"string"` with a `format` (regular expression) if necessary.  You can fill in the default values in your application logic.  If you really want the schema to drive the default values, you could do something like this:

```
{
  "oneOf":[
    {"type":"string"},
    {"enum":[ <your default values here> ]}
  ]
}
```

Secondly, if you allow custom values, how to you plan on validating (i.e. what are the "possible values" that a user could provide)?

**Comment by coderextreme** [comment]
I think they are only string values, but I can’t rule out other types.

I didn’t realize the oneOf worked like that, but I will try again.

We will not be validating custom values, only providing hints as to what to
values to use.

Thanks for the hint on regular expressions.

**Comment by coderextreme** [comment]
I have:
        "@name": {
            "oneOf": [
              {
                "enum": [
                  "accessRights",
                  "contributor",
                  "created",
                  "creator",
                  "description",
                  "drawing",
                  "error",
                  "generator",
                  "hint",
                  "identifier",
                  "Image",
                  "info",
                  "license",
                  "modified",
                  "MovingImage",
                  "photo",
                  "reference",
                  "requires",
                  "rights",
                  "robots",
                  "Sound",
                  "specificationSection",
                  "specificationUrl",
                  "subject",
                  "Text",
                  "title",
                  "TODO",
                  "translator",
                  "translated",
                  "version",
                  "warning"
                ]
              },
              {
                "type": "string"
              }
            ]
          },

Ajv, with some customization, reports:

keyword: oneOf
dataPath: X3D > head > meta > 0 > @name
value: "title"
message: should match exactly one schema in oneOf
params: {"passingSchemas":[0,1]}
file: C:/x3d-code/www.web3d.org/x3d/content/examples/X3dForWebAuthors/Chapter07EventAnimationInterpolation/NormalInterpolator.json
version: 3.3


Bug in Ajv?  Or in the property?  Should I put type first?

Thanks

**Comment by coderextreme** [comment]
json-schema 3.3 Validation error org.everit.json.schema.ValidationException: #/X3D/head/meta/24/@name: #: 2 subschemas matched instead of one C:/x3d-code/www.web3d.org/x3d/content/examples/X3dForWebAuthors/newScene.json

I see the problem.  I’m matching both subschemas!

Another suggestion? oneOrTwoOf ? severalOf ?

John

**Comment by handrews** [comment]
yeah you either need to change that `oneOf` to `anyOf`, or do some contortions to exclude the `enum` values from the `"type": "string"` branch which is more trouble than it's worth here.  `anyOf` should be fine.

**Comment by handrews** [comment]
This seems like the question was answered and no further comments for two weeks, so I'm closing it.  Please re-open if I am misinterpreting.
